docs/guide/options.md

## DVR-Scan Command-Line Reference

This page lists all DVR-Scan command line options grouped by functionality. Command line options always take precedence over values set with a [config file](config_file.md).


### General

 * <b><pre>-h, --help</pre></b>
    Show help message and all options and default values. Will also show any options overriden by your [user config file](config_file.md), if set.

 * <b><pre>-v, --version</pre></b>
    Show version information.

 * <b><pre>-L, --license</pre></b>
    Show copyright information.

The following options control terminal output:

 * <b><pre>--logfile log.txt</pre></b>
    File to save (append) log output.

 * <b><pre>-q, --quiet</pre></b>
    Suppress all console output except for final cutting list.

 * <b><pre>--verbosity</pre></b>
    Set verbosity of output messages, must be one of: <b><pre>debug, info, warning, error</pre></b>
    *Default*: `--verbosity info`


### Input/Output

 * <b><pre>-i video.mp4, --input video.mp4</pre></b>
    Path to input video. May specify multiple input videos so long as they have the same resolution and framerate. Wildcards/globs are supported (e.g. `-i folder/*.mp4`). Extracted motion events use the filename of the first video only as a prefix.

 * <b><pre>-c settings.cfg, --config settings.cfg</pre></b>
    Path to config file. If not set, tries to load one automatically from the user settings folder. See the [config file documentation](config_file.md) for details.

 * <b><pre>-d path, --output-dir path</pre></b> Write all output files to `path`. If not specified, files are written in the working directory.

 * <b><pre>-m mode, --output-mode mode</pre></b> Mode to use for saving motion events. Must be one of:

    * <b><pre>opencv</pre></b>
        (*Default*) Use OpenCV for saving motion events. Requires output to be in .AVI format.

    * <b><pre>ffmpeg</pre></b>
        Use ffmpeg for saving motion events. Ffmpeg encoder args can be set using a [config file](config_file.md#inputoutput) (option: *ffmpeg-output-args*). Does not work with overlays.

    * <b><pre>copy</pre></b>
        Same as `ffmpeg`, but using stream copying mode (no reencoding). Can potentially create inaccurate events due to keyframe placement.

 * <b><pre>-o video.avi, --output video.avi</pre></b> Save all motion events to a single file, instead of the default (one file per event). Only supported with the default output mode (`opencv`). Requires `.avi` extension.

 * <b><pre>-mo mask.avi, --mask-output mask.avi</pre></b> Save a video containing the calculated motion mask on each frame. Useful for tuning motion detection parameters. Requires `.avi` extension.

 * <b><pre>-so, --scan-only</pre></b> Do not save/extract events, only perform motion detection and display results.


### Seeking / Duration

All time values can be given as a timecode (`HH:MM:SS` or `HH:MM:SS.nnn`), in seconds as a number followed by `s` (`123s` or `123.45s`), or as number of frames (e.g. `1234`).

 * <b><pre>-st time, --start-time time</pre></b>Timecode in video to start motion detection from

 * <b><pre>-dt time, --duration time</pre></b> Maximum duration of input to process.

 * <b><pre>-et time, --end-time time</pre></b> Timecode to stop processing input.

For example, to scan the input between 1m30s and 2m:

    dvr-scan -i video.mp4 -st 00:01:30 -et 00:02:00


### Motion Events

All time values can be given as a timecode (`HH:MM:SS` or `HH:MM:SS.nnn`), in seconds as a number followed by `s` (`123s` or `123.45s`), or as number of frames (e.g. `1234`).

 * <b><pre>-l time, --min-event-length time</pre></b> Amount of time/frames that must have a motion score above the threshold setting before triggering a new event.
<br/>*Default*: `--min-event-length 2`

 * <b><pre>-tb time, --time-before-event time</pre></b> Maximum amount of time to include before each event.
<br/>*Default*: `--time-before-event 1.5s`

 * <b><pre>-tp time, --time-post-event time</pre></b> Maximum amount of time to include after each event. The event will end once no motion has been detected for this period of time.
<br/>*Default*: `--time-post-event time 2.0s`

For example, to save 10 seconds before or after any frame that exceeds the threshold:

    dvr-scan -i video.mp4 -l 1 -tb 10s -tp 10s


### Detection Parameters

When modifying these parameters, it can be useful to generate a motion mask (`-mo mask.avi`) to visually see how DVR-Scan processes the input when looking for motion events.

 * <b><pre>-b type, --bg-subtractor type</pre></b> The type of background subtractor to use. Must be one of:

    * <b><pre>MOG2</pre></b> (*Default*) [MOG2 Background Subtractor](https://docs.opencv.org/3.4/d7/d7b/classcv_1_1BackgroundSubtractorMOG2.html).

    * <b><pre>CNT</pre></b> [CNT Background Subtractor](https://docs.opencv.org/3.4/de/dca/classcv_1_1bgsegm_1_1BackgroundSubtractorCNT.html), faster than `MOG` but uses different method, so may need to adjust threshold/kernel size.

    * <b><pre>MOG2_CUDA</pre></b> [Nvidia CUDA-based version of MOG2](https://docs.opencv.org/3.4/df/d23/classcv_1_1cuda_1_1BackgroundSubtractorMOG2.html).


 * <b><pre>-t value, --threshold value</pre></b> Threshold representing the minimum amount of motion a frame must have to trigger an event. Lower values are more sensitive, requiring less movement. If the threshold is too high, some movement in the scene may not be detected. Setting the threshold too low can result in false detections. May need to be adjusted when modifying other parameters (e.g. `-b`/`--bg-subtractor` or `-k`/`--kernel-size`).
<br/>*Default*: `--threshold 0.15`

 * <b><pre>-k size, --kernel-size size</pre></b> Size in pixels of the noise reduction kernel. Must be an *odd* integer at least 3 or greater. If set to -1 (default), the kernel size will be calculated automatically based on input video resolution. Can be increased to filter out smaller movements from a scene. Default values: 7 for 1080p or greater, 5 for 720p, 3 for 480p.
<br/>*Default*: `--kernel-size -1`

Detection can be limited to a smaller region of the frame using the `-roi`/`--region-of-interest` flag:

 * <b><pre>-roi, --region-of-interest</pre></b> Show a pop-up window to select a region of interest using the mouse. The first frame will be displayed.

 * <b><pre>-roi x0, y0 w, h</pre></b> Rectangle specified as top-left corner and size. For example, `-roi 50,75, 100,150` specifies a 100x150 rectangle with the top left corner at coordinates (50,75). (0,0) is the top-left corner of the video.

 * <b><pre>-roi width,height</pre></b> Same as `-roi` but shrinks the window to fit within (width x height). Useful for processing videos larger than the monitor resolution.

The following options can improve performance, but may reduce detection accuracy:

 * <b><pre>-df factor, --downscale-factor factor</pre></b> Integer factor to downscale (shrink) video before processing, to improve performance. For example, if input video resolution is 1024 x 400, and factor=2, each frame is reduced to 1024/2 x 400/2=512 x 200 before processing.

 * <b><pre>-fs num_frames, --frame-skip num_frames</pre></b> Number of frames to skip after processing a given frame. Improves performance, at expense of frame and time accuracy, and may increase probability of missing motion events. If set, `-l`/`--min-event-length`, `-tb`/`--time-before-event`, and `-tp`/`--time-post-event` will all be scaled relative to the source framerate. Values above 1 or 2 are not recommended.
<br/><br/>When using the default output mode (`opencv`), skipped frames are not included. Set `-m`/`--output-mode` to `ffmpeg` or `copy` to include all frames from the input video when writing motion events to disk.
<br/><br/>Although adjusted for frame skipping, bounding box smoothing may be inconsistent when using frame skipping. Set `-bb 0` to disable smoothing.

### Overlays

 * <b><pre>-bb, --bounding-box</pre></b> Draw a bounding box around the areas where motion is detected.
<br/><br/>An optional amount of time for temporal smoothing can also be specified (e.g. `-bb 0.1s`). The default smoothing amount is 0.1 seconds. If set to 0 (`-bb 0`), smoothing is disabled.
<br/><br/>The color, thickness, and minimum size can be set with a [config file](config_file.md#bounding-box-overlay).

 * <b><pre>-tc, --time-code</pre></b>  Draw time code of each frame on the top left corner.
<br/><br/>Text properties (e.g. color, font size, margin) can be set with a [config file](config_file.md#text-overlays).

 * <b><pre>-fm, --frame-metrics</pre></b>  Draw frame metrics, including motion score, on the top right corner.
<br/><br/>Text properties (e.g. color, font size, margin) can be set with a [config file](config_file.md#text-overlays).

For example, to draw a bounding box and timecode on extracted motion events:

    dvr-scan -i video.mp4 -bb -tc

Overlays are only supported with the default output mode (`opencv`).


docs/changelog.md

DVR-Scan Changelog
==========================================================

## DVR-Scan 1.6

### 1.6 (In Development)

#### Release Notes

In development.
z
#### Changelog

**General**:

 - [enhancement] Noise reduction kernel can now be disabled by setting `-k`/`--kernel-size` to `0` or `1` ([#123](https://github.com/Breakthrough/DVR-Scan/issues/123))
 - [feature] New `-fm` / `--frame-metrics` option draws motion score on each frame to help tune [detection parameters](https://dvr-scan.readthedocs.io/en/latest/guide/options/#detection-parameters)
 - [other] Config file options that started with `time-code-` have been renamed to start with `text-`, and are now shared between the `time-code` and `frame-metrics` overlays:
    - `time-code-margin` is now `text-margin`
    - `time-code-font-scale` is now `text-font-scale`
    - `time-code-font-thickness` is now `text-font-thickness`
    - `time-code-font-color` is now `text-font-color`
    - `time-code-bg-color` is now `text-bg-color`


## DVR-Scan 1.5

### 1.5.1 (August 15, 2022) &nbsp;<span class="fa fa-tags"></span>

#### Changelog

 * [bugfix] Fix crash when opening multiple input videos ([#95](https://github.com/Breakthrough/DVR-Scan/issues/95))
 * [bugfix] Fix incorrect warning regarding frame decode failures at end of video


### 1.5 (July 30, 2022)

#### Release Notes

 * Significant performance improvements on multicore systems
 * Support wildcards/globs as inputs for scanning entire folders (`-i folder/*.mp4`)
 * Allow use of ffmpeg for better output quality (`-m ffmpeg`) or codec-copying mode (`-m copy`)
 * Configuration files are now supported, [see documentation for details](https://dvr-scan.readthedocs.io/en/v1.5/guide/config_file/)
     * Can specify config file path with `-c`/`--config`, or create a `dvr-scan.cfg` file in your user config folder
 * Windows binaries are now signed, thanks [SignPath.io](https://signpath.io/) (certificate by [SignPath Foundation](https://signpath.org/))
 * Experimental Nvidia CUDA® support has been added (set `-b MOG2_CUDA`)
    * If installing via `pip`, requires manual installation of OpenCV compiled with CUDA support
    * If downloading Windows version, make sure to download the GPU-enabled build (`dvr-scan-1.5-win64-cuda.zip`)
    * CUDA-enabled builds are not code signed, and do not include the `CNT` method
 * Minimum supported Python version is now 3.7
 * Minimum supported OpenCV version is now 3.x

#### Command-Line Changes

 * New options:
    * `-c`/`--config` - set path to config file
    * `-d`/`--output-dir` - set directory to write output files (default is working directory)
    * `-m`/`--output-mode` - set output mode (one of: `opencv`, `ffmpeg`, `copy`)
    * `-mo`/`--mask-output` - path to write motion mask for analysis
    * `--verbosity` and `--logfile` - control output verbosity and path to save output
 * `-i`/`--input` now supports globs/wildcards to scan entire folders, e.g. `-i folder/*.mp4`
 * Change default value for `-l`/`--min-event-length` to 0.1 seconds, previously was 2 frames
 * Long form of `-roi` has been renamed to `--region-of-interest` (previously was `--rectangle-of-interest`)
 * `-c` is now used for `--config`, previously was for `--codec`
 * Add experimental `MOG2_CUDA` option for `-b`/`--bg-subtractor`
 * Rename existing `MOG` option to `MOG2`
 * `--codec` has been removed, the value should now be set using a [config file](https://dvr-scan.readthedocs.io/en/v1.5/guide/config_file/)

#### Changelog

 * [feature] Configuration file support and new `-c`/`--config` argument to specify path to config files ([#77](https://github.com/Breakthrough/DVR-Scan/issues/77))
     * Breaks existing behavior of `-c` (was previously the shortform of `--codec`)
 * [feature] Add support for multiple output modes via `-m`/`--output-mode` argument ([#27](https://github.com/Breakthrough/DVR-Scan/issues/27), [#42](https://github.com/Breakthrough/DVR-Scan/issues/42))
 * [feature] Experimental support for GPU-based CUDA MOG2 filter ([#12](https://github.com/Breakthrough/DVR-Scan/issues/12))
 * [feature] Video encoding and decoding are now done in parallel with the scanning logic leading to improved performance on most systems ([#52](https://github.com/Breakthrough/DVR-Scan/issues/52))
 * [feature] Add support for exporting motion masks via `-mo`/`--mask-output` argument
     * Useful for detailed analysis or tuning of detection parameters
     * ffmpeg can be used to generate output videos by specifying `-m ffmpeg`
     * Codec-copy mode, using ffmpeg, can be used by specifying `-m copy`
 * [feature] Add `--verbosity` and `--logfile` arguments to provide more control over program output
 * [feature] Allow scanning entire folders using wildcards with `-i`/`--input` ([#5](https://github.com/Breakthrough/DVR-Scan/issues/5))
     * Glob expansion is also performed on each input path directly, so quoted globs also function correctly
 * [bugfix] Fix incorrect results when `-st`/`--start-time` is set
 * [bugfix] Event start times are now correctly calculated when using `-fs`/`--frame-skip` ([#68](https://github.com/Breakthrough/DVR-Scan/issues/68), [#70](https://github.com/Breakthrough/DVR-Scan/issues/70))
    * Note that all skipped frames within the event window are included in motion event, thus the calculated start time may be slightly earlier
 * [bugfix] Only get screen resolution when required (was causing issues on headless machines)
 * [bugfix] Fix output messages conflicting with progress bar shown during scanning
 * [bugfix] Output events now start from 1 to align with the event list
 * [bugfix] Event end times now include the presentation duration of the last frame

dvr_scan/cli.py
# -*- coding: utf-8 -*-
#
#      DVR-Scan: Video Motion Event Detection & Extraction Tool
#   --------------------------------------------------------------
#       [  Site: https://github.com/Breakthrough/DVR-Scan/   ]
#       [  Documentation: http://dvr-scan.readthedocs.org/   ]
#
# Copyright (C) 2014-2022 Brandon Castellano <http://www.bcastell.com>.
# PySceneDetect is licensed under the BSD 2-Clause License; see the
# included LICENSE file, or visit one of the above pages for details.
#
""" ``dvr_scan.cli`` Module

This module provides the get_cli_parser() function, which provides
an argparse-based CLI parser used by the DVR-Scan application.
"""

# Standard Library Imports
from __future__ import print_function
import argparse

# DVR-Scan Library Imports
import dvr_scan


def timecode_type_check(metavar=None):
    """ Creates an argparse type for a user-inputted timecode.

    The passed argument is declared valid if it meets one of three valid forms:
      1) Standard timecode; in form HH:MM:SS or HH:MM:SS.nnn
      2) Number of seconds; type # of seconds, followed by s (e.g. 54s, 0.001s)
      3) Exact number of frames; type # of frames (e.g. 54, 1000)
     valid integer which
    is greater than or equal to min_val, and if max_val is specified,
    less than or equal to max_val.

    Returns:
        A function which can be passed as an argument type, when calling
        add_argument on an ArgumentParser object

    Raises:
        ArgumentTypeError: Passed argument must be integer within proper range.
    """
    metavar = 'value' if metavar is None else metavar

    def _type_checker(value):
        valid = False
        value = str(value).lower().strip()
        # Integer number of frames.
        if value.isdigit():
            # All characters in string are digits, just parse as integer.
            frames = int(value)
            if frames >= 0:
                valid = True
                value = frames
        # Integer or real/floating-point number of seconds.
        elif value.endswith('s'):
            secs = value[:-1]
            if secs.replace('.', '').isdigit():
                secs = float(secs)
                if secs >= 0.0:
                    valid = True
                    value = secs
        # Timecode in HH:MM:SS[.nnn] format.
        elif ':' in value:
            tc_val = value.split(':')
            if (len(tc_val) == 3 and tc_val[0].isdigit() and tc_val[1].isdigit()
                    and tc_val[2].replace('.', '').isdigit()):
                hrs, mins = int(tc_val[0]), int(tc_val[1])
                secs = float(tc_val[2]) if '.' in tc_val[2] else int(tc_val[2])
                if (hrs >= 0 and mins >= 0 and secs >= 0 and mins < 60 and secs < 60):
                    valid = True
                    value = [hrs, mins, secs]
        if not valid:
            raise argparse.ArgumentTypeError(
                'invalid timecode: %s\n'
                'Timecode must be specified as number of frames (12345), seconds (number followed'
                ' by s, e.g. 123s or 123.45s), or timecode (HH:MM:SS[.nnn].' % value)
        return value

    return _type_checker


def int_type_check(min_val, max_val=None, metavar=None):
    """ Creates an argparse type for a range-limited integer.

    The passed argument is declared valid if it is a valid integer which
    is greater than or equal to min_val, and if max_val is specified,
    less than or equal to max_val.

    Returns:
        A function which can be passed as an argument type, when calling
        add_argument on an ArgumentParser object

    Raises:
        ArgumentTypeError: Passed argument must be integer within proper range.
    """
    metavar = 'value' if metavar is None else metavar

    def _type_checker(value):
        value = int(value)
        valid = True
        msg = ''
        if max_val is None:
            if value < min_val:
                valid = False
            msg = 'invalid choice: %d (%s must be at least %d)' % (value, metavar, min_val)
        else:
            if value < min_val or value > max_val:
                valid = False
            msg = 'invalid choice: %d (%s must be between %d and %d)' % (value, metavar, min_val,
                                                                         max_val)
        if not valid:
            raise argparse.ArgumentTypeError(msg)
        return value

    return _type_checker


def odd_int_type_check(min_val, max_val=None, metavar=None, allow_zero=True):
    """ Creates an argparse type for a range-limited integer which must be odd.

    The passed argument is declared valid if it is a valid integer which is odd
    (i.e. the modulus of the value with respect to two is non-zero), is greater
    than or equal to min_val, and, if specified, less than or equal to max_val.

    Returns:
        A function which can be passed as an argument type, when calling
        add_argument on an ArgumentParser object

    Raises:
        ArgumentTypeError: Argument must be odd integer within specified range.
    """
    metavar = 'value' if metavar is None else metavar

    def _type_checker(value):
        value = int(value)
        valid = True
        msg = ''
        if value == -1:
            return -1
        if value == 0 and allow_zero is True:
            return 0
        if (value % 2) == 0:
            valid = False
            msg = 'invalid choice: %d (%s must be an odd number)' % (value, metavar)
        elif max_val is None:
            if value < min_val:
                valid = False
            msg = 'invalid choice: %d (%s must be at least %d)' % (value, metavar, min_val)
        else:
            if value < min_val or value > max_val:
                valid = False
            msg = 'invalid choice: %d (%s must be between %d and %d)' % (value, metavar, min_val,
                                                                         max_val)
        if not valid:
            raise argparse.ArgumentTypeError(msg)
        return value

    return _type_checker


def float_type_check(min_val, max_val=None, metavar=None, default_str=None):
    """ Creates an argparse type for a range-limited float.

    The passed argument is declared valid if it is a valid float which is
    greater thanmin_val, and if max_val is specified, less than max_val.

    Returns:
        A function which can be passed as an argument type, when calling
        add_argument on an ArgumentParser object

    Raises:
        ArgumentTypeError: Passed argument must be float within proper range.
    """
    metavar = 'value' if metavar is None else metavar

    def _type_checker(value):
        if default_str and isinstance(value, str) and default_str == value:
            return None
        value = float(value)
        valid = True
        msg = ''
        if max_val is None:
            if value < min_val:
                valid = False
            msg = 'invalid choice: %3.1f (%s must be greater than %3.1f)' % (value, metavar,
                                                                             min_val)
        else:
            if value < min_val or value > max_val:
                valid = False
            msg = 'invalid choice: %3.1f (%s must be between %3.1f and %3.1f)' % (value, metavar,
                                                                                  min_val, max_val)
        if not valid:
            raise argparse.ArgumentTypeError(msg)
        return value

    return _type_checker


def string_type_check(valid_strings, case_sensitive=True, metavar=None):
    """ Creates an argparse type for a list of strings.

    The passed argument is declared valid if it is a valid string which exists
    in the passed list valid_strings.  If case_sensitive is False, all input
    strings and strings in valid_strings are processed as lowercase.  Leading
    and trailing whitespace is ignored in all strings.

    Returns:
        A function which can be passed as an argument type, when calling
        add_argument on an ArgumentParser object

    Raises:
        ArgumentTypeError: Passed argument must be string within valid list.
    """
    metavar = 'value' if metavar is None else metavar
    valid_strings = [x.strip() for x in valid_strings]
    if not case_sensitive:
        valid_strings = [x.lower() for x in valid_strings]

    def _type_checker(value):
        value = str(value)
        valid = True
        if not case_sensitive:
            value = value.lower()
        if not value in valid_strings:
            valid = False
            case_msg = ' (case sensitive)' if case_sensitive else ''
            msg = 'invalid choice: %s (valid settings for %s%s are: %s)' % (
                value, metavar, case_msg, valid_strings.__str__()[1:-1])
        if not valid:
            raise argparse.ArgumentTypeError(msg)
        return value

    return _type_checker


# pylint: disable=too-few-public-methods
class AboutAction(argparse.Action):
    """Custom argparse action for displaying the DVR-Scan ABOUT_STRING.

    Based off of the default VersionAction for displaying a string to the user.
    """

    # pylint: disable=redefined-builtin, too-many-arguments
    def __init__(self,
                 option_strings,
                 version=None,
                 dest=argparse.SUPPRESS,
                 default=argparse.SUPPRESS,
                 help="show version number and license/copyright information"):
        super(AboutAction, self).__init__(
            option_strings=option_strings, dest=dest, default=default, nargs=0, help=help)
        self.version = version

    def __call__(self, parser, namespace, values, option_string=None):
        version = self.version
        if version is None:
            version = parser.version
        parser.exit(message=version)


def get_cli_parser():
    """Creates the DVR-Scan argparse command-line interface.

    Returns:
        ArgumentParser object, which parse_args() can be called with.
    """
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    # pylint: disable=protected-access
    parser._optionals.title = 'arguments'

    parser.add_argument('-v', '--version', action=AboutAction, version=dvr_scan.ABOUT_STRING)

    parser.add_argument(
        '-i',
        '--input',
        metavar='VIDEO_FILE',
        required=True,
        type=str,
        action='append',
        help=('[REQUIRED] Path to input video. May be specified multiple'
              ' times to join several videos with the same resolution'
              ' and framerate. Any output filenames will be generated'
              ' using the first filename only.'))

    parser.add_argument(
        '-o',
        '--output',
        metavar='OUTPUT_VIDEO.avi',
        type=str,
        help=('If specified, all motion events will be written to a single'
              ' file, creating a compilation of only the frames in'
              ' the input video containing motion. By default each'
              ' motion event is written to a separate file. Filename'
              ' MUST end with .avi.'))

    parser.add_argument(
        '-b',
        '--bg-subtractor',
        metavar='TYPE',
        dest='bg_subtractor',
        type=string_type_check(['MOG', 'CNT'], False, 'TYPE'),
        default='MOG',
        help=('The type of background subtractor to use, must be one of: '
              ' MOG (default), CNT (parallel).'))

    parser.add_argument(
        '-so',
        '--scan-only',
        dest='scan_only_mode',
        action='store_true',
        default=False,
        help=('Only perform motion detection (does not write any files to disk).'))

    # TODO(v1.5): This needs to be changed to a new -c/--config flag.
    # Just leave default as XVID, since the other codecs don't seem to be as well supported,
    # and add a config file option to override it instead.
    parser.add_argument(
        '-c',
        '--codec',
        metavar='FOURCC',
        dest='fourcc_str',
        type=string_type_check(['XVID', 'MP4V', 'MP42', 'H264'], False, 'FOURCC'),
        default='XVID',
        help=('The four-letter identifier of the encoder/video codec to use'
              ' when exporting motion events as videos. Possible values'
              ' are: XVID, MP4V, MP42, H264.'))

    parser.add_argument(
        '-t',
        '--threshold',
        metavar='value',
        dest='threshold',
        type=float_type_check(0.0, None, 'value'),
        default=0.15,
        help=('Threshold value representing the amount of motion in a frame'
              ' required to trigger a motion event. Lower values require'
              ' less movement, and are more sensitive to motion. If the threshold'
              ' is too high, some movement in the scene may not be detected,'
              ' while too low of a threshold can trigger a false motion event.'))

    parser.add_argument(
        '-k',
        '--kernel-size',
        metavar='N',
        dest='kernel_size',
        type=odd_int_type_check(3, None, 'N', True),
        default=-1,
        help=('Size in pixels of the noise reduction kernel. Must be an odd'
              ' integer greater than 1, or set to -1 to auto-set based on'
              ' input video resolution (default). If the kernel size is set too'
              ' large, some movement in the scene may not be detected.'))

    parser.add_argument(
        '-l',
        '--min-event-length',
        metavar='T',
        dest='min_event_len',
        type=timecode_type_check('T'),
        default=2,
        help=('Number of frames that must exceed the threshold in a row to trigger'
              ' a new motion event, effectively setting a minimum event length.'
              ' Can also be specified as a timecode or # of seconds.'))

    parser.add_argument(
        '-tp',
        '--time-post-event',
        metavar='T',
        dest='time_post_event',
        type=timecode_type_check('T'),
        default='2s',
        help=('Number of frames to include after each motion event ends.'
              ' Any new motion events that occur in this period are'
              ' automatically joined with the current motion event.'
              ' Can also be specified as a timecode or # of seconds.'))

    parser.add_argument(
        '-tb',
        '--time-before-event',
        metavar='T',
        dest='time_pre_event',
        type=timecode_type_check('T'),
        default='1.5s',
        help=('Number of frames to include before a motion event is detected.'
              ' Can also be specified as a timecode or # of seconds.'))

    parser.add_argument(
        '-q',
        '--quiet',
        dest='quiet_mode',
        action='store_true',
        default=False,
        help=('Suppress all output except for final comma-separated list of motion events.'
              ' Useful for computing or piping output directly into other programs/scripts.'))

    parser.add_argument(
        '-st',
        '--start-time',
        metavar='time',
        dest='start_time',
        type=timecode_type_check('time'),
        default=None,
        help=('Time to seek to in video before performing detection. Can be'
              ' given in number of frames (12345), seconds (number followed'
              ' by s, e.g. 123s or 123.45s), or timecode (HH:MM:SS[.nnn]).'))

    parser.add_argument(
        '-dt',
        '--duration',
        metavar='time',
        dest='duration',
        type=timecode_type_check('time'),
        default=None,
        help=('Length of time in input video to limit motion detection to (see'
              ' -st for valid timecode formats). Overrides -et.'))

    parser.add_argument(
        '-et',
        '--end-time',
        metavar='time',
        dest='end_time',
        type=timecode_type_check('time'),
        default=None,
        help=('Timecode to stop motion detection at (see -st for valid'
              'timecode formats).'))

    parser.add_argument(
        '-df',
        '--downscale-factor',
        metavar='factor',
        dest='downscale_factor',
        type=int_type_check(1, None, 'factor'),
        default=1,
        help=('Factor to downscale (shrink) video before processing, to'
              ' improve performance. For example, if input video resolution'
              ' is 1024 x 400, and factor=2, each frame is reduced to'
              ' 1024/2 x 400/2=512 x 200 before processing.'))

    parser.add_argument(
        '-fs',
        '--frame-skip',
        metavar='num_frames',
        dest='frame_skip',
        type=int_type_check(0, None, 'num_frames'),
        default=0,
        help=('Number of frames to skip after processing a given frame.'
              ' Improves performance, at expense of frame and time accuracy,'
              ' and may increase probability of missing motion events.'
              ' If set, -l, -tb, and -tp will all be scaled relative to the source'
              ' framerate. Values above 1 or 2 are not recommended.'))

    parser.add_argument(
        '-tc',
        '--time-code',
        dest='draw_timecode',
        action='store_true',
        default=False,
        help=('Draw time code of each frame on the top left corner.'))

    parser.add_argument(
        '-roi',
        '--region-of-interest',
        dest='roi',
        metavar='x0 y0 w h',
        nargs='*',
        default=None,
        help=('If set, scan only in selected area, which is selected in a popup window'
              ' (select with mouse, then press enter).'
              ' Can also specify the window in terms of x/y/w/h.'
              ' Example for pop-up window: dvr-scan -i video.mp4 -roi '
              ' Example for predefined rectangle: dvr-scan -i video.mp4 -roi 100 110 50 50 '))

    parser.add_argument(
        '-bb',
        '--bounding-box',
        metavar='SMOOTH_TIME',
        dest='bounding_box',
        type=timecode_type_check('SMOOTH_TIME'),
        nargs='?',
        default=None,
        const='0.1s',
        help=('If set, draws a bounding box around the area where motion was detected. The amount'
              ' of temporal smoothing can be specified in either frames (12345) or seconds (number'
              ' followed by s, e.g. 123s or 123.45s). If omitted, defaults to 0.1s. If set to 0,'
              ' smoothing is disabled.'))

    # TODO(v1.5): Add a new -m/--output-mode flag to specify whether to use ffmpeg or the
    # OpenCV VideoWriter for output. Also will need to add some flags to specify the ffmpeg
    # arguments, and a new flag called --keep-temp-files for how concatenation has to work.

    return parser
